tests/unit/parcelset/put_parcels.cpp

```diff
@@ -0,0 +1,239 @@
+    results.reserve(numparcels_default);
+
+    // create parcels
+    std::vector<hpx::parcelset::parcel> parcels;
+    for (int i = 0; i != numparcels_default; ++i)
+    {
+        hpx::lcos::local::promise<double> p_arg;
+        hpx::lcos::promise<hpx::id_type> p_cont;
+
+        parcels.push_back(
+            generate_parcel<test2_action>(id, p_cont.get_id(),
+                p_arg.get_future())
+        );
+
+        args.push_back(std::move(p_arg));
+        results.push_back(p_cont.get_future());
+    }
+
+    // send parcels
+    hpx::get_runtime().get_parcel_handler().put_parcels(std::move(parcels));
+
+    // now make the futures ready
+    for (hpx::lcos::local::promise<double>& arg : args)
+    {
+        arg.set_value(42.0);
+    }
+
+    // verify all messages got actually sent to the correct locality
+    hpx::wait_all(results);
+
+    for (hpx::future<hpx::id_type>& f : results)
+    {
+        HPX_TEST(f.get() == id);
+    }
+}
+
+void test_mixed_arguments(hpx::id_type const& id)
+{
+    std::vector<double> data(vsize_default);
+    std::generate(data.begin(), data.end(), std::rand);
+
+    std::vector<hpx::lcos::local::promise<double> > args;
+    args.reserve(numparcels_default);
+
+    std::vector<hpx::future<hpx::id_type> > results;
+    results.reserve(numparcels_default);
+
+    // create parcels
+    std::vector<hpx::parcelset::parcel> parcels;
+    for (int i = 0; i != numparcels_default; ++i)
+    {
+        hpx::lcos::promise<hpx::id_type> p_cont;
+
+        if (std::rand() % 2)
+        {
+            parcels.push_back(
+                generate_parcel<test1_action>(id, p_cont.get_id(), data)
+            );
+        }
+        else
+        {
+            hpx::lcos::local::promise<double> p_arg;
+
+            parcels.push_back(
+                generate_parcel<test2_action>(id, p_cont.get_id(),
+                    p_arg.get_future())
+            );
+
+            args.push_back(std::move(p_arg));
+        }
+
+        results.push_back(p_cont.get_future());
+    }
+
+    // send parcels
+    hpx::get_runtime().get_parcel_handler().put_parcels(std::move(parcels));
+
+    // now make the futures ready
+    for (hpx::lcos::local::promise<double>& arg : args)
+    {
+        arg.set_value(42.0);
+    }
+
+    // verify all messages got actually sent to the correct locality
+    hpx::wait_all(results);
+
+    for (hpx::future<hpx::id_type>& f : results)
+    {
+        HPX_TEST(f.get() == id);
+    }
+}
+
+///////////////////////////////////////////////////////////////////////////////
+void print_counters(char const* name)
+{
+    using namespace hpx::performance_counters;
+
+    std::vector<performance_counter> counters = discover_counters(name);
+
+    for (performance_counter const& c : counters)
+    {
+        counter_value value = c.get_counter_value_sync();
+        hpx::cout
+            << "counter: " << c.get_name_sync()
+            << ", value: " << value.get_value<double>()
```

Comment:
On 01/07/2016 01:59 PM, Hartmut Kaiser wrote:
> In tests/unit/parcelset/put_parcels.cpp
> <https://github.com/STEllAR-GROUP/hpx/pull/1942#discussion_r49070149>:
> 
>> +    }
>> +}
>> +
>> +///////////////////////////////////////////////////////////////////////////////
>> +void print_counters(char const* name)
>> +{
>> +    using namespace hpx::performance_counters;
>> +
>> +    std::vector<performance_counter> counters = discover_counters(name);
>> +
>> +    for (performance_counter const& c : counters)
>> +    {
>> +        counter_value value = c.get_counter_value_sync();
>> +        hpx::cout
>> +            << "counter: " << c.get_name_sync()
>> +            << ", value: " << value.get_value<double>()
> 
> That could be done, I was not sure if this is really 100% reproducible.
> Printing the values was more for my sake.

I completely understand, it would be nice to have something that gets
automatically tested to ensure proper operation.